frontend/src/components/worksheets/Worksheet/Worksheet.js

```diff
@@ -1120,25 +1120,30 @@ class Worksheet extends React.Component {
             if (this.hasEditPermission()) {
                 var editor = ace.edit('worksheet-editor');
                 if (saveChanges) {
-                    this.setState({
-                        ws: {
-                            ...this.state.ws,
-                            info: {
-                                ...this.state.ws.info,
-                                source: editor.getValue().split('\n'),
+                    // Use callback function to ensure the worksheet info will not be sent to the backend until the frontend state has finished updating
+                    this.setState(
```

Comment:
Actually, can we wrap this in a promise and then use `await` instead? Given that we want to refactor callbacks into promises anyway

Reply:
Actually for setState, I do not think we should wrap it into a promise and use await since it already provides a callback to take actions when setState is completed

Reply:
What's the drawback of wrapping it in a promise and using await? It seems like it would make the code look at a bit cleaner, related to @Levilian 's work of changing callbacks -> promises?

something like this:

```
await new Promise(e => this.setState({...}, e));
```

If there's a drawback in doing it this way, we can keep using the callbacks, though.

Reply:
I updated the PR to use `promise` instead since you mentioned in the offline discussion that we have a code style convention of using `promise` instead.